<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>

<body>
    <script>
        // 继续暴露问题
        // 在新建实例的时候，在实例的回调函数里面添加一个定时器
        class Hd {
            static PENDING = "pending"
            static FULFILLED = "fulfilled"
            static REJECTED = "rejected"
            constructor(executor) {
                this.status = Hd.PENDING
                this.value = null
                try {
                    executor(this.resolve.bind(this),
                        this.reject.bind(this))
                } catch (error) {
                    this.reject(error)
                }
            }
            resolve(value) {
                if (this.status == Hd.PENDING) {
                    this.value = value
                    this.status = Hd.FULFILLED
                }
            }
            reject(reason) {
                if (this.status == Hd.PENDING) {
                    this.value = reason
                    this.status = Hd.REJECTED
                }
            }

            then(onFulfilled, onRejected) {
                console.log(1)
                if (!(onFulfilled instanceof Function)) {
                    onFulfilled = () => {}
                }
                if (!(onRejected instanceof Function)) {
                    onRejected = () => {}
                }

                // 刚刚进来的实例状态是pending
                // 而此时resolve还没执行，实例状态还没有变成pending
                // 所以虽然执行了then函数，但是不会执行then里的回调函数

                // 而原生的promise应该是，无论promise实例的resolve什么时候执行
                // 只要执行了resolve，就会调用then里的回调函数
                console.log(this.status)
                if (this.status == Hd.FULFILLED) {
                    setTimeout(() => {
                        try {
                            onFulfilled(this.value)
                        } catch (error) {
                            onRejected(error)
                        }
                    })
                }
                if (this.status == Hd.REJECTED) {
                    setTimeout(() => {
                        try {
                            onRejected(this.value)
                        } catch (error) {
                            onRejected(error) // 待定
                        }
                    })
                }

                // 两秒之后再看实例的状态
                setTimeout(() => {
                    console.log(this.status)
                }, 2000)
            }
        }

        // 在新建promise实例的时候，使用定时器延迟一秒改变状态
        // 结果then不会执行里面的回调函数
        let promise = new Hd((resolve, reject) => {
                setTimeout(() => {
                    console.log('执行resolve前的操作')
                    resolve('解决')
                }, 1000)

                // reject('拒绝')
            })
            .then((val) => {
                console.log(val)
            }, (reason) => {
                console.log(reason)
            })

        console.log('houdunren')
    </script>
</body>

</html>